algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -166,78 +267,90 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            let instance_combiners = &batch_combiners[&circuit.id].instance_combiners;
+            let _circuit_id_z_m = &circuit.id;
+            let _circuit_id_t = &circuit.id;
+            let circuit_constraint_domain = circuit_specific_state.constraint_domain;
+            let circuit_input_domain = circuit_specific_state.input_domain;
+            job_pool.add_job(move || {
+                let summed_z_m_poly_time =
+                    start_timer!(|| format!("Compute z_m poly for circuit {:?}", _circuit_id_z_m));
+                let summed_z_m = cfg_iter!(first_msg_i)
+                    .zip_eq(instance_combiners)
+                    .map(|(entry, combiner)| {
+                        let z_a = entry.z_a_poly.polynomial().as_dense().unwrap();
+                        let mut z_b = entry.z_b_poly.polynomial().as_dense().unwrap().clone();
+                        assert!(z_a.degree() < max_constraint_domain.size());
+                        if MM::ZK {
+                            assert_eq!(z_b.degree(), circuit_constraint_domain.size());
+                        } else {
+                            assert!(z_b.degree() < circuit_constraint_domain.size());
+                        }
+
+                        // we want to calculate batch_combiner_i * (z_a + eta_b * z_b + eta_c * z_a * z_b);
+                        // we rewrite this as  batch_combiner_i * (z_a * (eta_c * z_b + 1) + eta_b * z_b);
+                        // This is better since it reduces the number of required
+                        // multiplications by `circuit_constraint_domain.size()`.
+                        let mut summed_z_m = {
+                            // Mutate z_b in place to compute eta_c * z_b + 1
+                            // This saves us an additional memory allocation.
+                            cfg_iter_mut!(z_b.coeffs).for_each(|b| *b *= eta_c);
+                            z_b.coeffs[0] += F::one();
+                            let mut multiplier = PolyMultiplier::new();
+                            multiplier.add_polynomial_ref(z_a, "z_a");
+                            multiplier.add_polynomial_ref(&z_b, "eta_c_z_b_plus_one");
+                            multiplier.add_precomputation(fft_precomputation, ifft_precomputation);
+                            let result = multiplier.multiply().unwrap();
+                            // Start undoing in place mutation, by first subtracting the 1 that we added...
+                            z_b.coeffs[0] -= F::one();
+                            result
+                        };
+                        // ... and then multiplying by eta_b/eta_c, instead of just eta_b.
+                        cfg_iter_mut!(summed_z_m.coeffs).zip(&z_b.coeffs).for_each(|(c, b)| *c += eta_b_over_eta_c * b);
+
+                        // Multiply by linear combination coefficient.
+                        cfg_iter_mut!(summed_z_m.coeffs).for_each(|c| *c *= combiner);
```

Comment:
I wonder if it wouldn't be beneficial for performance if these 2 operations were to be combined with a
```
.for_each(|(c, b)| {
    *c += eta_b_over_eta_c * b;
    *c *= combiner;
})
```

Reply:
eta_b and summed_z_m have different lengths, so that doesn't work, unfortunately. (It's also why it's a `zip` there instead of a `zip_eq`.) I think there was a comment to that effect, but perhaps it got removed in a refactoring.

Reply:
3a5d5aac0